### ISSUE TITLE
Cells that get stuck running can sometimes not be stopped.

### ISSUE THREAD

**Issue by Treit** [dotnet/interactive#1456]
### Describe the bug
Cells that get stuck running can sometimes not be stopped by clicking the 'Stop cell' button.

I have seen multiple times a cell that gets stuck in a running state and clicking the little square 'stop cell' button does nothing.

I recently saw this with the following trivial notebook:

    #!csharp

    int i = 1234;

    #!javascript

    let x = 0
    interactive.csharp.getVariable('i').then(v => console.log(v))
    console.log(x)

    #!markdown

    <h1>Test!</h1>
    Test test test

For some reason this showed several cells that were 'stuck' running and clicking the stop button did nothing.

## .NET Interactive Version
Version: 1.0.230903+1a272ef3782d6e88b6d36ebd648fb1882da0a530
Build date: 2021-06-28T20:51:38.7685995Z

- OS
    - [x] Windows 11 (build 22000.51)
    - [ ] Windows 10
    - [ ] macOS
    - [ ] Linux (Please specify distro)
    - [ ] iOS
    - [ ] Android
 - Browser 
    - [ ] Chrome
    - [x] Edge
    - [ ] Firefox
    - [ ] Safari
 - Frontend
    - [ ] Jupyter Notebook
    - [ ] Jupyter Lab
    - [ ] nteract
    - [x] Visual Studio Code 
    - [ ] Visual Studio Code Insiders
    - [ ] Other (please specify)

### Screenshots
This animation shows the issue:

![vscode2](https://user-images.githubusercontent.com/8271812/123876941-1b335900-d8f1-11eb-991c-3eb2ce9454a6.gif)

**Comment by RandomFractals** [comment]
I hit this issue too after creating simple data pull C# cell in a .Net interactive .ipynb notebook.

It worked the first time I created it. Restarting and running that notebook in both latest stable and insiders either does nothing or gets still in infinite cell run loop. 

You can try my test from this ticket: https://github.com/RandomFractals/vscode-data-table/issues/26#issuecomment-875798847

**Comment by jonsequitur** [comment]
Is the code snippet in the initial comment relevant to this bug? That looks the same as the code presented in #1452 (which I can repro) but the screen shot seems to be showing something different (which I cannot repro). 

I'm using version v1.0.2351010 of the extension.

**Comment by RandomFractals** [comment]
@jonsequitur I don't think it really matters what code you run. I ran into when testing your JS cells before and can't run that C# notebook we created yesterday either. Using your latest ext. v. too.

There is some disconnect in either vscode notebooks UI/kernel plumbing or something else. 

Is there a debug we can turn on for your extension to see what messages it gets and what it is doing?

**Comment by jonsequitur** [comment]
You can always open the developer web view tools to see if there are errors.

Do you have a link to the C# notebook that's not working for you?

**Comment by RandomFractals** [comment]
same notebook from yesterday I referenced in my issue above. tried in vscode v1.58.0 today too: https://github.com/RandomFractals/vscode-data-table/issues/26#issuecomment-876767189

https://github.com/RandomFractals/vscode-data-table/blob/main/notebooks/chicago-speed-cameras.ipynb

![dib-data-notebook](https://user-images.githubusercontent.com/656833/124994976-482bee00-e00c-11eb-9cc1-89193a53f047.png)

**Comment by RandomFractals** [comment]
is it possible that this is related to #1464? although my latest test notebook doesn't use JS.

Here are the errors and traces I get in console dev tools:

![image](https://user-images.githubusercontent.com/656833/125075026-b4960400-e083-11eb-8257-5f0ee0898186.png)

I had it run to 757 getHttpAPIEndpoint Command calls, which I belive come from your ext./kernel, and waited for over 300 seconds.

The only way to stop it is close that notebook. This is with your latest ext. v. using vscode vf1.58.0.

Let me know if you want me to enable verbose diagnostics for more info, but I believe this should get you started for now.

Also, why does your extension go through new install every time I open .net .ipynb notebook?

**Comment by jonsequitur** [comment]
The variable viewer doesn't work with .NET Interactive yet. You can use the `#!whos` magic command to see your current top-level variables.

**Comment by brettfo** [comment]
@RandomFractals Can you click View -> Output then change the drop down to ".NET Interactive : diagnostics" and copy/paste the contents?  It'll list the port that the HTTP component is listening on and I'd like to see the values.  E.g., the contents of that output channel on my machine look like:

```
Extension started for VS Code Stable.
Kernel for '*DOTNET-INTERACTIVE-NOTEBOOK-SERIALIZATION*' started (6916).
Kernel process 6916 Port 64890 is using tunnel uri http://localhost:64890/
Kernel for 'untitled:Untitled-1.dib?dotnet-interactive' started (18104).
Kernel process 18104 Port 64895 is using tunnel uri http://localhost:64895/
```

**Comment by RandomFractals** [comment]
@bretto, so I got insiders v1.59.0 and your v1.0.2357011 udpate. Ran that notebook and got this in your diagnostics output. I think we are are getting closer on this. Thanks for letting me know how to run them:
```
Extension started for VS Code Insiders.
Kernel for 'file:///d%3A/dev/vscode/projects/vscode-data-table/notebooks/chicago-speed-cameras.ipynb' started (9276).
kernel (9276) stderr: Unhandled exception: 
kernel (9276) stderr: System.TypeInitializationException: The type initializer for 'System.Management.Automation.Runspaces.RunspaceFactory' threw an exception.
 ---> System.DllNotFoundException: Unable to load DLL 'api-ms-win-eventing-provider-l1-1-0.dll' or one of its dependencies: The specified module could not be found. (0x8007007E)
   at System.Management.Automation.Tracing.EtwActivity.UnsafeNativeMethods.EventActivityIdControl(ActivityControlCode controlCode, Guid& activityId)
   at System.Management.Automation.Tracing.EtwActivity.GetActivityId()
   at System.Management.Automation.Runspaces.RunspaceFactory..cctor()
   --- End of inner exception stack trace ---
   at System.Management.Automation.Runspaces.RunspaceFactory.CreateRunspace(PSHost host, InitialSessionState initialSessionState)
   at Microsoft.DotNet.Interactive.PowerShell.PowerShellKernel.CreatePowerShell() in D:\workspace\_work\1\s\src\Microsoft.DotNet.Interactive.PowerShell\PowerShellKernel.cs:line 98
   at System.Lazy`1.ViaFactory(LazyThreadSafetyMode mode)
   at System.Lazy`1.ExecutionAndPublication(LazyHelper executionAndPublication, Boolean useDefaultConstructor)
   at System.Lazy`1.CreateValue()
   at Microsoft.DotNet.Interactive.PowerShell.PowerShellKernel.SetVariableAsync(String name, Object value, Type declaredType) in D:\workspace\_work\1\s\src\Microsoft.DotNet.Interactive.PowerShell\PowerShellKernel.cs:line 164
   at Microsoft.DotNet.Interactive.VSCode.KernelExtensions.UseVSCodeHelpersAsync[TKernel](TKernel kernel, Kernel rootKernel) in D:\workspace\_work\1\s\src\Microsoft.DotNet.Interactive.VSCode\KernelExtensions.cs:line 17
   at Microsoft.DotNet.Interactive.App.CommandLine.CommandLineParser.<>c__DisplayClass4_5.<<Create>b__25>d.MoveNext() in D:\workspace\_work\1\s\src\dotnet-interactive\CommandLine\CommandLineParser.cs:line 426
--- End of stack trace from previous location ---
   at Microsoft.DotNet.Interactive.KernelExtensions.VisitSubkernelsAsync(Kernel kernel, Func`2 onVisit, Boolean recursive) in D:\workspace\_work\1\s\src\Microsoft.DotNet.Interactive\KernelExtensions.cs:line 387
   at Microsoft.DotNet.Interactive.App.CommandLine.CommandLineParser.<>c__DisplayClass4_0.<<Create>b__23>d.MoveNext() in D:\workspace\_work\1\s\src\dotnet-interactive\CommandLine\CommandLineParser.cs:line 421
--- End of stack trace from previous location ---
   at System.CommandLine.Invocation.CommandHandler.GetExitCodeAsync(Object value, InvocationContext context)
   at System.CommandLine.Invocation.ModelBindingCommandHandler.InvokeAsync(InvocationContext context)
   at System.CommandLine.Invocation.InvocationPipeline.<>c__DisplayClass4_0.<<BuildInvocationChain>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass23_0.<<UseParseErrorReporting>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at Microsoft.DotNet.Interactive.App.CommandLine.CommandLineParser.<>c__DisplayClass4_0.<<Create>b__3>d.MoveNext() in D:\workspace\_work\1\s\src\dotnet-interactive\CommandLine\CommandLineParser.cs:line 191
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass16_0.<<UseHelp>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass27_0.<<UseVersionOption>b__1>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass25_0.<<UseTypoCorrections>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseSuggestDirective>b__24_0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass22_0.<<UseParseDirective>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass11_0.<<UseDebugDirective>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<RegisterWithDotnetSuggest>b__10_0>d.MoveNext()
--- End of stack trace from previous l
kernel (9276) stderr: ocation ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass14_0.<<UseExceptionHandler>b__0>d.MoveNext()

Kernel for 'file:///d%3A/dev/vscode/projects/vscode-data-table/notebooks/chicago-speed-cameras.ipynb' ended (9276) with code 1.
```

**Comment by brettfo** [comment]
The 5.0.301 SDK is fine, it's what I have on my local machine.  I'll see if I can figure out what's going on with `api-ms-win-....`

**Comment by jonsequitur** [comment]
The exception is being thrown from the PowerShell kernel. It's a bug that this code path is being called if you're not using PowerShell, so I think fixing that will at least tactically unblock this.

**Comment by brettfo** [comment]
@jonsequitur The PowerShell kernel is getting invoked via SetVariableAsync as it tries to set the `$InteractiveHost` object.  We could do the tactical fix by making set variable an explicit command and adding it to the deferred list, instead of setting it outright.  I think this is ultimately where we'll go with this.

**Comment by jonsequitur** [comment]
You can use this to schedule a deferred command that's not intended to cross process boundaries: https://github.com/dotnet/interactive/blob/b12671692f472e9be9fd722a68339c1365048f79/src/Microsoft.DotNet.Interactive/Commands/AnonymousKernelCommand.cs#L6.

**Comment by RandomFractals** [comment]
why does it take 26 secs to run this?

![Screen Shot 2021-07-27 at 15 10 05](https://user-images.githubusercontent.com/656833/127528250-aa161ced-0d4d-49c0-b7f0-5be9c0c5233e.png)

that json data result loads in under 1 second in other notebook types.

**Comment by RandomFractals** [comment]
also, should not the result of JsonDocument.Parse(data) be in json format with `application/json` mime type?

Looks like you are generating html document for that last call too:

![Screen Shot 2021-07-29 at 13 30 51](https://user-images.githubusercontent.com/656833/127533243-6ff7fcdd-2cf9-4187-a08b-809ca02f6f82.png)

**Comment by RandomFractals** [comment]
on the bright side, I can confirm .net interactive notebooks run just fine in Code Spaces :slightly_smiling_face: 

![image](https://user-images.githubusercontent.com/656833/127739866-0b9fc915-76e9-4967-a34a-ed38e16230ea.png)

**Comment by boyestrous** [comment]
Howdy! I'm now seeing a very similar set of problems with newer versions of everything. Did anyone ever figure out a root cause for this issue?

To be clear: my c# notebook gets stuck when running certain cells in vscode. I've occasionally been able to get these same cells to run without issue and the code runs fine outside of the notebook environment. The stop button does not work and I have to kill processes and/or close the notebook to be able to continue working.

_note_: I seem to be able to run other notebook cells while the one cell is stuck. 
_note 2_: Sometimes the cell will actually finish running at some point. A simple for loop took ~15 minutes to complete, but it eventually did stop.

**Versions**
- Vscode: 1.72.2
- .Net Runtime: 6.0.402
- .Net Interactive Notebooks Extension: v1.0.3504060
